proposed/0006-unified-structured-data-imporvement.md

```diff
@@ -0,0 +1,147 @@
+
+##Expected outcome
+
+Removing the field `previous_owner_keys` simplifies code and reduces logical complexity during Structured Data Updates and ownership-transfers. This might further simplify vault logic in future to avoid Structured Data Collisions even for the same `location` (64 byte network address for the data) as vaults will now have to look only for a single field (`owners`) and track its changes to make the storage key unique even though `location` wasn't unique.
+
+# Detailed design
+
+##StructuredData
+
+```
+struct StructuredData {
+    tag_type   : u64,
+    identifier : crypto::hash::sha512::Digest,
+    version    : u64,                         // mutable - incrementing (deterministic) version number
+    data       : Vec<u8>,                     // mutable - in many cases this is encrypted
+    owner_keys : vec<crypto::sign::PublicKey> // mutable - n * 32 Bytes (where n is number of owners)
+    signatures : Vec<crypto::sign::Signature> // mutable - detached signature of all the `mutable` fields (barring itself) by owners
+}
+```
+
+Fixed (immutable fields)
+- tag_type
+- identifier
+
+##Flow
+
+The following shows Creation, Updation and Transfer-of-ownership of Structured Data:
+
+**step 0:**
+Client-A sends:
+```
+StructuredData {
+    other_fields: Value-00,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 1:**
+Vault receives for the first time. So Just Stores it:
+```
+StructuredData {
+    other_fields: Value-00,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 2:**
+Client-A updates it:
+```
+StructuredData {
+    other_fields: Value-11,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 3:**
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 1 to verify signature of data in step 2. If valid replace previous data and store new one:
+```
+StructuredData {
+    other_fields: Value-11,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 4:**
+Client-A wants to transfer ownership of Structured Data to Client-B:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 5:**
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 3 to verify signature of data in step 4. If valid replace previous data and store new one:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 6:**
+Client-A tries to update illegally:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 7:**
+Vault will use owner field from **step 5** == `B-PublicKey` and signature verification fails. So it will not update anything.
```

Comment:
> Then even though the PublicKeys in the owner fields would match, Signature still has to be checked and then rejected as illegal update. 

Yeah, I forgot to mention that apart from StructuredData, the real message also contains the PublicKey of the client together with signature of this message, routing will check this signature during parsing of the message so by the time the StructuredData reaches Vault/Client, you can be sure that it came from the given client.

> So signature check is the ultimate indicator of a legal/illegal update.

So, yes, but it happens implicitly in routing.